# TV Script Generation

In this project, you'll generate your own [Seinfeld](https://en.wikipedia.org/wiki/Seinfeld) TV scripts using RNNs.  You'll be using part of the [Seinfeld dataset](https://www.kaggle.com/thec03u5/seinfeld-chronicles#scripts.csv) of scripts from 9 seasons.  The Neural Network you'll build will generate a new ,"fake" TV script, based on patterns it recognizes in this training data.

## Get the Data

The data is already provided for you in `./data/Seinfeld_Scripts.txt` and you're encouraged to open that file and look at the text. 
>* As a first step, we'll load in this data and look at some samples. 
* Then, you'll be tasked with defining and training an RNN to generate a new script!

In [1]:
# Want to be able to use traceback
import os
os.environ['CUDA_LAUNCH_BLOCKING'] = "1"

In [2]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# load in data
import helper
data_dir = './data/Seinfeld_Scripts.txt'
text = helper.load_data(data_dir)

## Explore the Data
Play around with `view_line_range` to view different parts of the data. This will give you a sense of the data you'll be working with. You can see, for example, that it is all lowercase text, and each new line of dialogue is separated by a newline character `\n`.

In [3]:
view_line_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in text.split()})))

lines = text.split('\n')
print('Number of lines: {}'.format(len(lines)))
word_count_line = [len(line.split()) for line in lines]
print('Average number of words in each line: {}'.format(np.average(word_count_line)))

print()
print('The lines {} to {}:'.format(*view_line_range))
print('\n'.join(text.split('\n')[view_line_range[0]:view_line_range[1]]))

Dataset Stats
Roughly the number of unique words: 46367
Number of lines: 109233
Average number of words in each line: 5.544240293684143

The lines 0 to 10:
jerry: do you know what this is all about? do you know, why were here? to be out, this is out...and out is one of the single most enjoyable experiences of life. people...did you ever hear people talking about we should go out? this is what theyre talking about...this whole thing, were all out now, no one is home. not one person here is home, were all out! there are people trying to find us, they dont know where we are. (on an imaginary phone) did you ring?, i cant find him. where did he go? he didnt tell me where he was going. he must have gone out. you wanna go out you get ready, you pick out the clothes, right? you take the shower, you get all ready, get the cash, get your friends, the car, the spot, the reservation...then youre standing around, what do you do? you go we gotta be getting back. once youre out, you wanna get back! y

---
## Implement Pre-processing Functions
The first thing to do to any dataset is pre-processing.  Implement the following pre-processing functions below:
- Lookup Table
- Tokenize Punctuation

### Lookup Table
To create a word embedding, you first need to transform the words to ids.  In this function, create two dictionaries:
- Dictionary to go from the words to an id, we'll call `vocab_to_int`
- Dictionary to go from the id to word, we'll call `int_to_vocab`

Return these dictionaries in the following **tuple** `(vocab_to_int, int_to_vocab)`

In [4]:
import problem_unittests as tests

def create_lookup_tables(text):
    """
    Create lookup tables for vocabulary
    :param text: The text of tv scripts split into words
    :return: A tuple of dicts (vocab_to_int, int_to_vocab)
    """
    # TODO: Implement Function
    # encode the text and map each character to an integer and vice versa

    # we create two dictionaries:
    # 1. int2char, which maps integers to characters
    # 2. char2int, which maps characters to unique integers
    chars = tuple(set(text)) # no duplicates wanted
    int_to_vocab = dict(enumerate(chars)) #dict that maps each char to unique integer
    vocab_to_int = {word: i for i, word in enumerate(chars, 0)}

    # encode the text
    encoded = np.array([vocab_to_int[ch] for ch in text])
    
    print(int_to_vocab)
    print(' ')
    print(vocab_to_int)
    print(' ')
    print(encoded)
    
    # return tuple
    return (vocab_to_int, int_to_vocab)


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_create_lookup_tables(create_lookup_tables)

{0: 'matter', 1: 'puke', 2: 'rotch', 3: 'on', 4: 'check', 5: 'give', 6: "moe's", 7: 'homer_simpson', 8: 'and', 9: 'ice', 10: 'you', 11: 'self', 12: 'hey', 13: 'of', 14: "i'm", 15: 'days', 16: 'your', 17: 'lately', 18: 'these', 19: 'homer', 20: 'not', 21: 'eh', 22: 'meet', 23: 'normal', 24: 'problems', 25: 'barney_gumble', 26: "i'll", 27: 'another', 28: 'last', 29: 'hello', 30: 'there', 31: 'gonna', 32: 'catch', 33: 'whats', 34: 'hold', 35: 'to', 36: 'anybody', 37: 'little', 38: 'pick', 39: 'tavern', 40: "you're", 41: 'forget', 42: 'moe_szyslak', 43: 'bart_simpson', 44: 'the', 45: 'seen', 46: 'my', 47: 'back', 48: 'where', 49: 'got', 50: 'enhance', 51: 'yeah', 52: 'i', 53: 'has', 54: 'listen', 55: 'one', 56: 'skills', 57: 'is', 58: 'mike', 59: 'name', 60: 'carve', 61: 'moe', 62: 'with', 63: 'me', 64: 'effervescent', 65: 'should', 66: 'drink', 67: 'only', 68: 'social', 69: 'an', 70: 'elite'}
 
{'matter': 0, 'puke': 1, 'rotch': 2, 'on': 3, 'check': 4, 'give': 5, "moe's": 6, 'homer_simpson

### Tokenize Punctuation
We'll be splitting the script into a word array using spaces as delimiters.  However, punctuations like periods and exclamation marks can create multiple ids for the same word. For example, "bye" and "bye!" would generate two different word ids.

Implement the function `token_lookup` to return a dict that will be used to tokenize symbols like "!" into "||Exclamation_Mark||".  Create a dictionary for the following symbols where the symbol is the key and value is the token:
- Period ( **.** )
- Comma ( **,** )
- Quotation Mark ( **"** )
- Semicolon ( **;** )
- Exclamation mark ( **!** )
- Question mark ( **?** )
- Left Parentheses ( **(** )
- Right Parentheses ( **)** )
- Dash ( **-** )
- Return ( **\n** )

This dictionary will be used to tokenize the symbols and add the delimiter (space) around it.  This separates each symbols as its own word, making it easier for the neural network to predict the next word. Make sure you don't use a value that could be confused as a word; for example, instead of using the value "dash", try using something like "||dash||".

In [5]:
def token_lookup():
    """
    Generate a dict to turn punctuation into a token.
    :return: Tokenized dictionary where the key is the punctuation and the value is the token
    """
    # TODO: Implement Function
    values = ['<PERIOD>', '<COMMA>', '<QUOTATION_MARK>', '<SEMICOLON>', '<EXCLAMATION_MARK>', 
              '<QUESTION_MARK>', '<LEFT_PAREN>', '<RIGHT_PAREN>', '<QUESTION_MARK>', 
              '<NEW_LINE>', '<DASH>']
    keys = ['.', ',', '"', ';', '!', '?', '(', ')', '?', '\n', '-']
    token = dict(zip(keys, values)) 
    
    #print(token)
    
    return token

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_tokenize(token_lookup)

Tests Passed


## Pre-process all the data and save it

Running the code cell below will pre-process all the data and save it to file. You're encouraged to lok at the code for `preprocess_and_save_data` in the `helpers.py` file to see what it's doing in detail, but you do not need to change this code.

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# pre-process training data
helper.preprocess_and_save_data(data_dir, token_lookup, create_lookup_tables)

{0: 'liberated', 1: "stick'n", 2: 'dokey', 3: 'ehm', 4: 'whaddaya', 5: "'bit'", 6: 'accident', 7: 'opportunity', 8: "'reer'", 9: 'goop', 10: 'crestfallen', 11: 'abutment', 12: 'doves', 13: 'surrounded', 14: 'revolting', 15: 'kest', 16: 'shaver', 17: 'penny', 18: 'b', 19: 'pharaohs', 20: '84th', 21: 'hospital', 22: 'presses', 23: 'tully', 24: "jehova's", 25: 'surrendered', 26: '*jerk*', 27: 'clumps', 28: 'canola', 29: 'machay', 30: 'intervention', 31: 'mon', 32: 'ruthie', 33: 'starches', 34: 'hal:', 35: 'craw', 36: 'laserium', 37: 'fredo', 38: '<buyers>', 39: 'exports', 40: 'procreate', 41: 'brake', 42: 'credits', 43: 'nightingale', 44: 'warns', 45: "evening's", 46: 'fan', 47: 'wife', 48: 'sawdust', 49: 'ooooo', 50: 'pepperoni', 51: 'out}', 52: 'ez', 53: "'fantastic", 54: 'shelters', 55: 'intrigues', 56: 'blower', 57: 'directory', 58: 'capricorn', 59: "incident'", 60: 'tioga', 61: 'ultrasound', 62: 'learned', 63: '100', 64: 'im', 65: 'strangled', 66: 'debonair', 67: 'awrghh', 68: 'marac

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [7]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

int_text, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()

## Build the Neural Network
In this section, you'll build the components necessary to build an RNN by implementing the RNN Module and forward and backpropagation functions.

### Check Access to GPU

In [8]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import torch

# Check for a GPU
train_on_gpu = torch.cuda.is_available()
if not train_on_gpu:
    print('No GPU found. Please use a GPU to train your neural network.')

## Input
Let's start with the preprocessed input data. We'll use [TensorDataset](http://pytorch.org/docs/master/data.html#torch.utils.data.TensorDataset) to provide a known format to our dataset; in combination with [DataLoader](http://pytorch.org/docs/master/data.html#torch.utils.data.DataLoader), it will handle batching, shuffling, and other dataset iteration functions.

You can create data with TensorDataset by passing in feature and target tensors. Then create a DataLoader as usual.
```
data = TensorDataset(feature_tensors, target_tensors)
data_loader = torch.utils.data.DataLoader(data, 
                                          batch_size=batch_size)
```

### Batching
Implement the `batch_data` function to batch `words` data into chunks of size `batch_size` using the `TensorDataset` and `DataLoader` classes.

>You can batch words using the DataLoader, but it will be up to you to create `feature_tensors` and `target_tensors` of the correct size and content for a given `sequence_length`.

For example, say we have these as input:
```
words = [1, 2, 3, 4, 5, 6, 7]
sequence_length = 4
```

Your first `feature_tensor` should contain the values:
```
[1, 2, 3, 4]
```
And the corresponding `target_tensor` should just be the next "word"/tokenized word value:
```
5
```
This should continue with the second `feature_tensor`, `target_tensor` being:
```
[2, 3, 4, 5]  # features
6             # target
```

Updated batching code from https://knowledge.udacity.com/questions/29798

In [9]:
from torch.utils.data import TensorDataset, DataLoader
def batch_data(words, sequence_length, batch_size):
    """
    Batch the neural network data using DataLoader
    :param words: The word ids of the TV scripts
    :param sequence_length: The sequence length of each batch
    :param batch_size: The size of each batch; the number of sequences in a batch
    :return: DataLoader with batched data
    """
    # TODO: Implement function
    # total number of words in a batch
    batch_size_total = batch_size * sequence_length
    # total number of batches that can be made
    #n_batches = len(words) // batch_size_total
    # keep only enough words to make full batches
    dwords = words[:-sequence_length]
    
    # transform words into a numpy array for reshaping
    #words = np.array(words)
    # reshape into batch_size rows
    #words = words.reshape((batch_size, -1))
    feature_tensors = []
    target_tensors = []
    # iterate through the array, one sequence at a time
    for n in range(0, len(dwords)):
        feature_tensors.append(words[n:n+sequence_length])
        target_tensors.append(words[n+sequence_length])
    
    nbr_batch = len(dwords)//batch_size
    feature_tensors = feature_tensors[:nbr_batch*batch_size]
    target_tensors = target_tensors[:nbr_batch*batch_size]
    # transform from numpy arrays into torch tensors
    torch_features = torch.from_numpy(np.array(feature_tensors))
    torch_targets = torch.from_numpy(np.array(target_tensors))
   
    
    # create and batch data
    data = TensorDataset(torch_features, torch_targets)
    data_loader = torch.utils.data.DataLoader(data, batch_size=batch_size, shuffle=True)
    
    # return a dataloader
    return data_loader

### Test your dataloader 

You'll have to modify this code to test a batching function, but it should look fairly similar.

Below, we're generating some test text data and defining a dataloader using the function you defined, above. Then, we are getting some sample batch of inputs `sample_x` and targets `sample_y` from our dataloader.

Your code should return something like the following (likely in a different order, if you shuffled your data):

```
torch.Size([10, 5])
tensor([[ 28,  29,  30,  31,  32],
        [ 21,  22,  23,  24,  25],
        [ 17,  18,  19,  20,  21],
        [ 34,  35,  36,  37,  38],
        [ 11,  12,  13,  14,  15],
        [ 23,  24,  25,  26,  27],
        [  6,   7,   8,   9,  10],
        [ 38,  39,  40,  41,  42],
        [ 25,  26,  27,  28,  29],
        [  7,   8,   9,  10,  11]])

torch.Size([10])
tensor([ 33,  26,  22,  39,  16,  28,  11,  43,  30,  12])
```

### Sizes
Your sample_x should be of size `(batch_size, sequence_length)` or (10, 5) in this case and sample_y should just have one dimension: batch_size (10). 

### Values

You should also notice that the targets, sample_y, are the *next* value in the ordered test_text data. So, for an input sequence `[ 28,  29,  30,  31,  32]` that ends with the value `32`, the corresponding output should be `33`.

In [10]:
# test dataloader

test_text = range(50)
t_loader = batch_data(test_text, sequence_length=5, batch_size=10)

data_iter = iter(t_loader)
sample_x, sample_y = data_iter.next()

print(sample_x.shape)
print(sample_x)
print()
print(sample_y.shape)
print(sample_y)

torch.Size([10, 5])
tensor([[ 21,  22,  23,  24,  25],
        [ 28,  29,  30,  31,  32],
        [ 13,  14,  15,  16,  17],
        [ 27,  28,  29,  30,  31],
        [ 11,  12,  13,  14,  15],
        [  5,   6,   7,   8,   9],
        [ 31,  32,  33,  34,  35],
        [  0,   1,   2,   3,   4],
        [  4,   5,   6,   7,   8],
        [ 37,  38,  39,  40,  41]])

torch.Size([10])
tensor([ 26,  33,  18,  32,  16,  10,  36,   5,   9,  42])


---
## Build the Neural Network
Implement an RNN using PyTorch's [Module class](http://pytorch.org/docs/master/nn.html#torch.nn.Module). You may choose to use a GRU or an LSTM. To complete the RNN, you'll have to implement the following functions for the class:
 - `__init__` - The initialize function. 
 - `init_hidden` - The initialization function for an LSTM/GRU hidden state
 - `forward` - Forward propagation function.
 
The initialize function should create the layers of the neural network and save them to the class. The forward propagation function will use these layers to run forward propagation and generate an output and a hidden state.

**The output of this model should be the *last* batch of word scores** after a complete sequence has been processed. That is, for each input sequence of words, we only want to output the word scores for a single, most likely, next word.

### Hints

1. Make sure to stack the outputs of the lstm to pass to your fully-connected layer, you can do this with `lstm_output = lstm_output.contiguous().view(-1, self.hidden_dim)`
2. You can get the last batch of word scores by shaping the output of the final, fully-connected layer like so:

```
# reshape into (batch_size, seq_length, output_size)
output = output.view(batch_size, -1, self.output_size)
# get last batch
out = output[:, -1]
```

https://github.com/warmspringwinds/pytorch-rnn-sequence-generation-classification/blob/master/notebooks/unconditional_lyrics_training.ipynbhttps://github.com/warmspringwinds/pytorch-rnn-sequence-generation-classification/blob/master/notebooks/unconditional_lyrics_training.ipynb

https://github.com/kaustav1987/TV-Script-generation/blob/master/Seinfeld_TV_script_generation_using_LSTM.ipynb for a lot of the bug fixes and trouble shooting. This was helpful with the OOP part since this student seemed to have a really good understanding of the ".self" operations.


In [11]:
import torch.nn as nn

class RNN(nn.Module):
    
    def __init__(self, vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.5):
        """
        Initialize the PyTorch RNN Module
        :param vocab_size: The number of input dimensions of the neural network (the size of the vocabulary)
        :param output_size: The number of output dimensions of the neural network
        :param embedding_dim: The size of embeddings, should you choose to use them        
        :param hidden_dim: The size of the hidden layer outputs
        :param dropout: dropout to add in between LSTM/GRU layers
        """
        super(RNN, self).__init__()
        
        self.vocab_size = vocab_size
        self.output_size = output_size
        self.embedding_dim = embedding_dim
        self.hidden_dim = hidden_dim
        self.n_layers = n_layers
        self.dropout = dropout
         
        # One-hot encoding function    
        self.embedding = nn.Embedding(self.vocab_size, self.embedding_dim)

        # define model layers
        self.lstm = nn.LSTM(self.embedding_dim, self.hidden_dim, self.n_layers, 
                            dropout=dropout, batch_first=True)
        
        # define dropout
        self.dropout = nn.Dropout(self.dropout)
        
        # define linear fc layer
        self.fc = nn.Linear(self.hidden_dim, self.output_size)

    
    def forward(self, nn_input, hidden):
        """
        Forward propagation of the neural network
        :param nn_input: The input to the neural network
        :param hidden: The hidden state        
        :return: Two Tensors, the output of the neural network and the latest hidden state
        """
        
        
        ### Converts labels into one-hot encoding and runs a linear 
        ### layer on each of the converted one-hot encoded elements
        #print(nn_input.shape)
        
        # get first element in batch
        batch_size = nn_input.shape[0]
        #print(batch_size)
        
        # convert to long tensor
        nn_input = nn_input.type(torch.cuda.LongTensor)
        #print(nn_input)
        
        # get the embedding from the input index
        nn_input = self.embedding(nn_input)
        #print(nn_input)
        
        # run batch through LSTM model, producing output and hidden state
        lstm_out, hidden = self.lstm(nn_input, hidden)
        
        # Flatten hidden state without changing the order of the targets (contiguous)
        lstm_out = lstm_out.contiguous().view(batch_size, -1, self.hidden_dim)
        #print(lstm_out)
        
        # dropout 
        lstm_out = self.dropout(lstm_out)
        
        # Fully connected layer
        lstm_out = self.fc(lstm_out)
        
        # get last batch
        lstm_out = lstm_out[:, -1]
        #print(lstm_out, hidden)
        
        # return one batch of output word scores and the hidden state
        return lstm_out, hidden
    
    
    def init_hidden(self, batch_size):
        '''
        Initialize the hidden state of an LSTM/GRU
        :param batch_size: The batch_size of the hidden state
        :return: hidden state of dims (n_layers, batch_size, hidden_dim)
        '''
        # initialize hidden state with zero weights, and move to GPU if available
        weight = next(self.parameters()).data

        if (train_on_gpu):
            hidden = (weight.new(self.n_layers, batch_size, self.hidden_dim).zero_().cuda(),
                  weight.new(self.n_layers, batch_size, self.hidden_dim).zero_().cuda())
        else:
            hidden = (weight.new(self.n_layers, batch_size, self.hidden_dim).zero_(),
                      weight.new(self.n_layers, batch_size, self.hidden_dim).zero_())

        return hidden

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_rnn(RNN, train_on_gpu)

Tests Passed


### Define forward and backpropagation

Use the RNN class you implemented to apply forward and back propagation. This function will be called, iteratively, in the training loop as follows:
```
loss = forward_back_prop(decoder, decoder_optimizer, criterion, inp, target)
```

And it should return the average loss over a batch and the hidden state returned by a call to `RNN(inp, hidden)`. Recall that you can get this loss by computing it, as usual, and calling `loss.item()`.

**If a GPU is available, you should move your data to that GPU device, here.**

Some mispellings/syntax fixes from https://github.com/thomasxmeng/Project-tv-script-generation-RNN/blob/master/dlnd_tv_script_generation_7.ipynb

In [12]:
def forward_back_prop(rnn, optimizer, criterion, inp, target, hidden):
    """
    Forward and backward propagation on the neural network
    :param decoder: The PyTorch Module that holds the neural network
    :param decoder_optimizer: The PyTorch optimizer for the neural network
    :param criterion: The PyTorch loss function
    :param inp: A batch of input to the neural network
    :param target: The target output for the batch of input
    :return: The loss and the latest hidden state Tensor
    """

    # move data to GPU, if available
    if(train_on_gpu):
        rnn.cuda()
        inp, target = inp.cuda(), target.cuda()
    
    ### perform backpropagation and optimization
    # create new vars for the hidden state, don't backprop entire history
    hidden = tuple([each.data for each in hidden])
    
    # zero accumulated gradients
    rnn.zero_grad()
    
    # get the output from the model 
    output, hidden = rnn(inp, hidden)
    #print(output, hidden)
    
    # Calculate loss and performn backprop
    loss = criterion(output, target)
    loss.backward()
    
    # clip_grad-Norm helps prevent the exploding gradient problem
    nn.utils.clip_grad_norm_(rnn.parameters(), 5)
    optimizer.step()
    loss = loss.item()
    
    # return the loss over a batch and the hidden state produced by our model
    return loss, hidden

# Note that these tests aren't completely extensive.
# they are here to act as general checks on the expected outputs of your functions
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_forward_back_prop(RNN, forward_back_prop, train_on_gpu)

Tests Passed


## Neural Network Training

With the structure of the network complete and data ready to be fed in the neural network, it's time to train it.

### Train Loop

The training loop is implemented for you in the `train_decoder` function. This function will train the network over all the batches for the number of epochs given. The model progress will be shown every number of batches. This number is set with the `show_every_n_batches` parameter. You'll set this parameter along with other parameters in the next section.

In [13]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""

def train_rnn(rnn, batch_size, optimizer, criterion, n_epochs, show_every_n_batches=100):
    batch_losses = []
    
    rnn.train()

    print("Training for %d epoch(s)..." % n_epochs)
    for epoch_i in range(1, n_epochs + 1):
        
        # initialize hidden state
        hidden = rnn.init_hidden(batch_size)
        
        for batch_i, (inputs, labels) in enumerate(train_loader, 1):
            
            # make sure you iterate over completely full batches, only
            n_batches = len(train_loader.dataset)//batch_size
            if(batch_i > n_batches):
                break
            
            # forward, back prop
            loss, hidden = forward_back_prop(rnn, optimizer, criterion, inputs, labels, hidden)          
            # record loss
            batch_losses.append(loss)

            # printing loss stats
            if batch_i % show_every_n_batches == 0:
                print('Epoch: {:>4}/{:<4}  Loss: {}\n'.format(
                    epoch_i, n_epochs, np.average(batch_losses)))
                batch_losses = []

    # returns a trained rnn
    return rnn

### Hyperparameters

Set and train the neural network with the following parameters:
- Set `sequence_length` to the length of a sequence.
- Set `batch_size` to the batch size.
- Set `num_epochs` to the number of epochs to train for.
- Set `learning_rate` to the learning rate for an Adam optimizer.
- Set `vocab_size` to the number of uniqe tokens in our vocabulary.
- Set `output_size` to the desired size of the output.
- Set `embedding_dim` to the embedding dimension; smaller than the vocab_size.
- Set `hidden_dim` to the hidden dimension of your RNN.
- Set `n_layers` to the number of layers/cells in your RNN.
- Set `show_every_n_batches` to the number of batches at which the neural network should print progress.

If the network isn't getting the desired results, tweak these parameters and/or the layers in the `RNN` class.

In [18]:
# Data params
# Sequence Length
sequence_length = 10
# Batch Size
batch_size = 256

# data loader - do not change
train_loader = batch_data(int_text, sequence_length, batch_size)

In [19]:
# Training parameters
# Number of Epochs
num_epochs = 5
# Learning Rate
learning_rate = .001

# Model parameters
# Vocab size
vocab_size = len(vocab_to_int)
# Output size
output_size = vocab_size
# Embedding Dimension
embedding_dim = 500
# Hidden Dimension
hidden_dim = 500
# Number of RNN Layers
n_layers = 2

# Show stats for every n number of batches
show_every_n_batches = 1000

print("sequence_length: ", sequence_length)
print("batch_size: ", batch_size)
print("learning_rate: ", learning_rate)
print("vocab_size: ", vocab_size)
print("output_size: ", output_size)
print("embedding_dim: ", embedding_dim)
print("hidden_dim: ", hidden_dim)
print("n_layers: ", n_layers)

sequence_length:  10
batch_size:  256
learning_rate:  0.001
vocab_size:  21388
output_size:  21388
embedding_dim:  500
hidden_dim:  500
n_layers:  2


### Train
In the next cell, you'll train the neural network on the pre-processed data.  If you have a hard time getting a good loss, you may consider changing your hyperparameters. In general, you may get better results with larger hidden and n_layer dimensions, but larger models take a longer time to train. 
> **You should aim for a loss less than 3.5.** 

You should also experiment with different sequence lengths, which determine the size of the long range dependencies that a model can learn.

In [20]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np


# create model and move to gpu if available
rnn = RNN(vocab_size, output_size, embedding_dim, hidden_dim, n_layers, dropout=0.2)
if train_on_gpu:
    rnn.cuda()

# defining loss and optimization functions for training
optimizer = torch.optim.Adam(rnn.parameters(), lr=learning_rate)
criterion = nn.CrossEntropyLoss()

# training the model
trained_rnn = train_rnn(rnn, batch_size, optimizer, criterion, num_epochs, show_every_n_batches)

# saving the trained model
helper.save_model('./save/trained_rnn', trained_rnn)
print('Model Trained and Saved')

Training for 5 epoch(s)...
Epoch:    1/5     Loss: 4.793264881134033

Epoch:    1/5     Loss: 4.273202593803406

Epoch:    1/5     Loss: 4.117586159467697

Epoch:    2/5     Loss: 3.9497232933571396

Epoch:    2/5     Loss: 3.86398380446434

Epoch:    2/5     Loss: 3.841763946533203

Epoch:    3/5     Loss: 3.7096505304873792

Epoch:    3/5     Loss: 3.642281055212021

Epoch:    3/5     Loss: 3.644327782869339

Epoch:    4/5     Loss: 3.511963009352311

Epoch:    4/5     Loss: 3.4800390787124633

Epoch:    4/5     Loss: 3.4911965613365172

Epoch:    5/5     Loss: 3.360710951195894

Epoch:    5/5     Loss: 3.32130712890625

Epoch:    5/5     Loss: 3.360844569683075



/opt/conda/lib/python3.6/site-packages/torch/serialization.py:193: UserWarning: Couldn't retrieve source code for container of type RNN. It won't be checked for correctness upon loading.
  "type " + obj.__name__ + ". It won't be checked "


Model Trained and Saved


### Question: How did you decide on your model hyperparameters? 
For example, did you try different sequence_lengths and find that one size made the model converge faster? What about your hidden_dim and n_layers; how did you decide on those?

**Answer:** 

 - Tried small batch size first, and had issues with the GPU memory
 - Tried learning rate of .01 but loss increased with each iteration (overshooting)
 - Increased embedding dimensions and hidden dimensions. This greatly helped the loss
 - Changed vocab size to the length of the vocab dictionary
 
**Batch Size**: 
 - Recommended starting values --> 1,2,4,8,16,32,64,128,256
 - Large: Takes more memory for training process. Computation boost.
 - Small: More noise in error calculations, helps training process stopping at local minimum. 
 - Increasing batch size helped the most
 - Yoshua Bengio in his 2012 paper "Practical Recommendations for Gradient-Based Training of Deep Architectures" recommends a default of 32
 

**Number of Hidden Units / Layers**:
 - Want number hidden units large enough
 - More compex, the higher it needs to be, larger usually doesn't hurt according to Bengio
 - "Larger than optimal values typically do not hurt generalization performance much, but of course they require proportionally more computation (if scaling all the layers atthe same time in a fully connected architecture)." p.11
 - According to Bengio, you could also do a layer-wise opimization of hyperparameters (maybe next time)
 - Without increasing batch size, adding another layer did little

**RNN Hyperparameters**
 - LSTM more commonly used
 - Layers: depth of at least 2. 3 layers could give mixed results.
 - Larger embedding better, but will stop optimizing at a certain size

---
# Checkpoint

After running the above training cell, your model will be saved by name, `trained_rnn`, and if you save your notebook progress, **you can pause here and come back to this code at another time**. You can resume your progress by running the next cell, which will load in our word:id dictionaries _and_ load in your saved model by name!

In [21]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import torch
import helper
import problem_unittests as tests

_, vocab_to_int, int_to_vocab, token_dict = helper.load_preprocess()
trained_rnn = helper.load_model('./save/trained_rnn')

## Generate TV Script
With the network trained and saved, you'll use it to generate a new, "fake" Seinfeld TV script in this section.

### Generate Text
To generate the text, the network needs to start with a single word and repeat its predictions until it reaches a set length. You'll be using the `generate` function to do this. It takes a word id to start with, `prime_id`, and generates a set length of text, `predict_len`. Also note that it uses topk sampling to introduce some randomness in choosing the most likely next word, given an output set of word scores!

In [22]:
"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
import torch.nn.functional as F

def generate(rnn, prime_id, int_to_vocab, token_dict, pad_value, predict_len=100):
    """
    Generate text using the neural network
    :param decoder: The PyTorch Module that holds the trained neural network
    :param prime_id: The word id to start the first prediction
    :param int_to_vocab: Dict of word id keys to word values
    :param token_dict: Dict of puncuation tokens keys to puncuation values
    :param pad_value: The value used to pad a sequence
    :param predict_len: The length of text to generate
    :return: The generated text
    """
    rnn.eval()
    
    # create a sequence (batch_size=1) with the prime_id
    current_seq = np.full((1, sequence_length), pad_value)
    current_seq[-1][-1] = prime_id
    predicted = [int_to_vocab[prime_id]]
    
    for _ in range(predict_len):
        if train_on_gpu:
            current_seq = torch.LongTensor(current_seq).cuda()
        else:
            current_seq = torch.LongTensor(current_seq)
        
        # initialize the hidden state
        hidden = rnn.init_hidden(current_seq.size(0))
        
        # get the output of the rnn
        output, _ = rnn(current_seq, hidden)
        
        # get the next word probabilities
        p = F.softmax(output, dim=1).data
        if(train_on_gpu):
            p = p.cpu() # move to cpu
         
        # use top_k sampling to get the index of the next word
        top_k = 5
        p, top_i = p.topk(top_k)
        top_i = top_i.numpy().squeeze()
        
        # select the likely next word index with some element of randomness
        p = p.numpy().squeeze()
        word_i = np.random.choice(top_i, p=p/p.sum())
        
        # retrieve that word from the dictionary
        word = int_to_vocab[word_i]
        predicted.append(word)     
        
        # the generated word becomes the next "current sequence" and the cycle can continue
        current_seq = np.roll(current_seq, -1, 1)
        current_seq[-1][-1] = word_i
    
    gen_sentences = ' '.join(predicted)
    
    # Replace punctuation tokens
    for key, token in token_dict.items():
        ending = ' ' if key in ['\n', '(', '"'] else ''
        gen_sentences = gen_sentences.replace(' ' + token.lower(), key)
    gen_sentences = gen_sentences.replace('\n ', '\n')
    gen_sentences = gen_sentences.replace('( ', '(')
    
    # return all the sentences
    return gen_sentences

### Generate a New Script
It's time to generate the text. Set `gen_length` to the length of TV script you want to generate and set `prime_word` to one of the following to start the prediction:
- "jerry"
- "elaine"
- "george"
- "kramer"

You can set the prime word to _any word_ in our dictionary, but it's best to start with a name for generating a TV script. (You can also start with any other names you find in the original text file!)

In [23]:
# run the cell multiple times to get different results!
gen_length = 400 # modify the length to your preference
prime_word = 'jerry' # name for starting the script

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
pad_word = helper.SPECIAL_WORDS['PADDING']
generated_script = generate(trained_rnn, vocab_to_int[prime_word + ':'], int_to_vocab, token_dict, vocab_to_int[pad_word], gen_length)
print(generated_script)

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:63: UserWarning: RNN module weights are not part of single contiguous chunk of memory. This means they need to be compacted at every call, possibly greatly increasing memory usage. To compact weights again call flatten_parameters().


jerry: shrimp' comforting.

newman: oh!

jerry:(to jerry) hey, you know, i'm just going to have the same exact same outfit.

elaine: i mean, what about the guy?

george:(to elaine) hey, you got that fedex?

jerry: yeah.

george: i don't know, but i don't have to go back to work.

elaine: oh, that's not my fault!

jerry: well, you know, you know, you want to be the same guy.

jerry: oh my god!

jerry:(to jerry) oh, yeah, well, we should get a little nervous about this.

elaine: oh, no no no no no no no no...

jerry: oh, you know, i just wanted to see what i am.

elaine: what are you gonna do?

elaine: oh, i was trying to see her name for a while.

george:(laughs, pointing at george) hey!

jerry: hey.

jerry: hey.

jerry: hey, hey!

george: hey, hey, hey.

elaine:(on the phone in a doorway) oh, i don't think so. i don't know...

jerry: i know.

kramer: well, i guess i can get a little uncomfortable.

jerry: oh, i was.

newman: i think it was the only one i had.(george is shocked, and loo

#### Save your favorite scripts

Once you have a script that you like (or find interesting), save it to a text file!

In [24]:
# save script to a text file
f =  open("generated_script_1.txt","w")
f.write(generated_script)
f.close()

# The TV Script is Not Perfect
It's ok if the TV script doesn't make perfect sense. It should look like alternating lines of dialogue, here is one such example of a few generated lines.

### Example generated script

>jerry: what about me?
>
>jerry: i don't have to wait.
>
>kramer:(to the sales table)
>
>elaine:(to jerry) hey, look at this, i'm a good doctor.
>
>newman:(to elaine) you think i have no idea of this...
>
>elaine: oh, you better take the phone, and he was a little nervous.
>
>kramer:(to the phone) hey, hey, jerry, i don't want to be a little bit.(to kramer and jerry) you can't.
>
>jerry: oh, yeah. i don't even know, i know.
>
>jerry:(to the phone) oh, i know.
>
>kramer:(laughing) you know...(to jerry) you don't know.

You can see that there are multiple characters that say (somewhat) complete sentences, but it doesn't have to be perfect! It takes quite a while to get good results, and often, you'll have to use a smaller vocabulary (and discard uncommon words), or get more data.  The Seinfeld dataset is about 3.4 MB, which is big enough for our purposes; for script generation you'll want more than 1 MB of text, generally. 

# Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_tv_script_generation.ipynb" and save another copy as an HTML file by clicking "File" -> "Download as.."->"html". Include the "helper.py" and "problem_unittests.py" files in your submission. Once you download these files, compress them into one zip file for submission.